src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+
+                    // Determine if this method is "invokable" in an analysis (can't be overridden)
+                    final boolean invokable = !this.currentClassIsExternallyExtendable() ||
+                        TreeInfo.isConstructor(tree) ||
+                        (tree.mods.flags & (Flags.STATIC | Flags.PRIVATE | Flags.FINAL)) != 0;
+
+                    // Add method or constructor to map
+                    final MethodInfo info = new MethodInfo(this.currentClass, tree, analyzable, invokable);
+                    ThisEscapeAnalyzer.this.methodMap.put(tree.sym, info);
+
+                    // Recurse
+                    super.visitMethodDef(tree);
+                } finally {
+                    this.lint = lintPrev;
+                }
+            }
+
+            // Determines if the current class could be extended in some external compilation unit
+            private boolean currentClassIsExternallyExtendable() {
+                return !this.currentClass.sym.isFinal() &&
+                  !(this.currentClass.sym.isSealed() && this.currentClass.permitting.isEmpty()) &&
+                  !(this.currentClass.sym.owner.kind == MTH) &&
+                  !this.privateOuter;
+            }
+        }.scan(tree);
+
+        // TODO: eliminate sealed classes where all permitted subclasses are in this compilation unit
+
+        // Now analyze all of the analyzable constructors we found
+        for (Map.Entry<Symbol, MethodInfo> entry : this.methodMap.entrySet()) {
+
+            // We are looking for analyzable constructors only
+            final Symbol sym = entry.getKey();
+            final MethodInfo methodInfo = entry.getValue();
+            if (!methodInfo.isAnalyzable())
+                continue;
+
+            // Analyze constructor body
+            this.targetClass = methodInfo.getDeclaringClass();
+            this.methodClass = this.targetClass;
+            Assert.check(this.depth == -1);
+            Assert.check(this.refs == null);
+            this.pushScope();
+            try {
+
+                // Add the initial 'this' reference
+                this.refs = RefSet.newEmpty();
+                this.refs.add(ThisRef.direct());
+
+                // Scan constructor statements
+                this.analyzeStatements(methodInfo.getDeclaration().body.stats);
+            } finally {
+                this.popScope();
+                this.methodClass = null;
+                this.targetClass = null;
+                this.refs = null;
+            }
+        }
+
+        // Eliminate duplicate warnings. Warning B duplicates warning A if the stack trace of A is a prefix
+        // of the stack trace of B. For example, if constructor Foo(int x) has a leak, and constructor
+        // Foo() invokes this(0), then emitting a warning for Foo() would be redundant.
+        final BiPredicate<DiagnosticPosition[], DiagnosticPosition[]> extendsAsPrefix = (warning1, warning2) -> {
```

Comment:
Another strategy would be to give a single warning per leaking constructor.

Reply:
We already impose a limit of at most one warning per constructor.

But for this limit what is technically meant by "per constructor" is "At most one warning per constructor, assuming that constructor is the one being analyzed (i.e., the one invoked by the subclass and not via `this()`)."

So that limitation doesn't stop a constructor from generating the same warning multiple times due to it being invoked indirectly by other constructors via `this()`. Those duplicate warnings are what the code above eliminates.

So this code only generates one warning, and it's reported for the second constructor:
```java
public class DupWarn1 {

    public DupWarn1() {
        this(0);
    }

    public DupWarn1(int x) {
        this.mightLeak();
    }

    public void mightLeak() {
    }
}

DupWarn1.java:8: warning: [this-escape] possible 'this' escape before subclass is fully initialized
        this.mightLeak();
                      ^
```
This is appropriate. The leak is really in the second constructor; the first constructor has nothing to do with it. Reporting a leak for both constructors would be redundant.

An interesting side question: Is it possible to come up with an example where constructor A has a 'this' leak, and some other constructor B invokes `this()` to delegate to A, but when B delegates to A the leak occurs differently, or not at all?

Because of the "flood" analysis, and the fact that you can't pass 'this' references to `this()` or `super()` (these are static contexts), I don't think it's possible.

So in fact the deduplication will always apply whenever `this()` is involved.